Added Swerve Drive Controller Package

nitin2606 opened the pull request:
Description
This PR introduces the swerve_drive_controller, a new controller for swerve drive robots with four independently steerable wheels, enabling omnidirectional motion in ROS 2. It complements controllers like diff_drive_controller by supporting advanced mobile robot platforms.

Features

Supports geometry_msgs/msg/Twist or TwistStamped velocity inputs (x, y linear; z angular).
Publishes raw nav_msgs/msg/Odometry for user-defined post-processing.
Publishes /tf transforms (optional, if enable_odom_tf=true).
Configurable via YAML for wheel geometry and kinematic constraints.
Changes
Added swerve_drive_controller package (src/, include/, test/).
Exported as a pluginlib plugin (swerve_drive_controller_plugin.xml).
Updated ros2_controllers CMakeLists.txt and package.xml.
Included gtest tests (test/test_swerve_drive_controller.cpp) and config (test/config/test_swerve_drive_controller.yaml).

nitin2606 committed:
Add swerve_drive_controller package with tests and documentation

nitin2606 committed:
Merge branch 'ros-controls:master' into feature/swerve-drive-controller

christophfroehlich commented on swerve_drive_controller/LICENSE:
no need to include the license file here as we have it in the top level of the repo, right?

Amronos replied:
You have to remove this file.

christophfroehlich commented on swerve_drive_controller/doc/userdoc.rst:
```diff
@@ -0,0 +1,47 @@
+
+Parameters
+,,,,,,,,,,,
+
```
Let's use GPL to render the docs

```suggestion
This controller uses the `generate_parameter_library <https://github.com/PickNikRobotics/generate_parameter_library>`_ to handle its parameters. The parameter `definition file located in the src folder <https://github.com/ros-controls/ros2_controllers/blob/{REPOS_FILE_BRANCH}/swerve_drive_controller/src/swerve_drive_controller_parameters.yaml>`_ contains descriptions for all the parameters used by the controller.

.. generate_parameter_library_details::
  ../src/swerve_drive_controller_parameters.yaml

An example parameter file could be
```

christophfroehlich commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,192 @@
+// Copyright 2025 (your name or organization)
```
please fill out, if you don't have an affiliation just add ros2_control development team

christophfroehlich commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_kinematics.hpp:
```diff
@@ -0,0 +1,99 @@
+// Copyright 2025 (your name or organization)
```
fill me

christophfroehlich commented on swerve_drive_controller/package.xml:
```diff
@@ -0,0 +1,40 @@
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
+  <buildtool_depend>ament_cmake_ros</buildtool_depend>
+  <!-- <build_depend>generate_parameter_library</build_depend> -->
```
why uncommented? should be of `<depend>` level
https://github.com/PickNikRobotics/generate_parameter_library?tab=readme-ov-file#add-parameter-library-generation-to-project

christophfroehlich commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,743 @@
+// Copyright 2025 (your name or organization)
```
fill me

christophfroehlich commented on swerve_drive_controller/src/swerve_drive_kinematics.cpp:
```diff
@@ -0,0 +1,108 @@
+// Copyright 2025 (your name or organization)
```
fill me

christophfroehlich commented on swerve_drive_controller/test/test_swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,58 @@
+// Copyright 2025 (your name or organization)
```
fill me

christophfroehlich commented on swerve_drive_controller/test/test_swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,457 @@
+// Copyright 2025 (your name or organization)
```
fill me

christophfroehlich reviewed (changes requested):
Thanks again for your contribution. Could you please fix the failing tests? A very quick first review below.

Do you mind add a brief section in the [kinematics section](https://github.com/ros-controls/ros2_controllers/blob/master/doc/mobile_robot_kinematics.rst) of the docs, even in a new PR to get it merged faster? Thanks!

nitin2606 commented on swerve_drive_controller/CMakeLists.txt:
```diff
@@ -0,0 +1,98 @@
+)
+
+target_include_directories(swerve_drive_controller PRIVATE include)
+ament_target_dependencies(swerve_drive_controller
+  builtin_interfaces
+  controller_interface
+  geometry_msgs
+  hardware_interface
+  nav_msgs
+  pluginlib
+  rclcpp
+  rclcpp_lifecycle
+  realtime_tools
+  tf2
+  tf2_msgs
+)
```
I'll refactor it.

nitin2606 commented:
I'll make changes and push.

nitin2606 committed:
Added Swerve Drive In mobile_robot_kinematics.rst and fixed CMakeLists.txt of swerve_drive_controller package

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Merge branch 'ros-controls:master' into feature/swerve-drive-controller

christophfroehlich commented:
do you mind cherry-picking the changes to the docs in a separate PR? Splitting this up would help to reviewing things.

nitin2606 committed:
Merge branch 'ros-controls:master' into feature/swerve-drive-controller

nitin2606 committed:
Added Swerve Drive Kinematics in mobile_robot_kinematics.rst

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 commented:
> do you mind cherry-picking the changes to the docs in a separate PR? Splitting this up would help to reviewing things.

I have raised a separate PR for changes in docs. PR #1712

bmagyar committed:
Merge branch 'master' into feature/swerve-drive-controller

bmagyar committed:
Apply suggestions from code review

nitin2606 commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,192 @@
+  std::shared_ptr<Wheel> get_wheel(const std::string & wheel_name);
+  std::shared_ptr<Axle> get_axle(const std::string & axle_name);
+
+  // Handles for three wheels and their axles
```
It's by mistake, it should be four wheels

nitin2606 commented on swerve_drive_controller/src/swerve_drive_kinematics.cpp:
```diff
@@ -0,0 +1,105 @@
+      wheel_positions_[i].second * wheel_positions_[i].second);
+  }
+  double wz_robot = wz_sum / wz_denominator;
+  // double wz_robot = wz_sum / 4.0;
```
Sure, we can remove the commented code

nitin2606 commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,740 @@
+  double y_offset = wheel_params_.y_offset;
+  double radius = wheel_params_.radius;
+
+  auto wheel_command_ =
```
Sure, it can be removed.

nitin2606 commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,192 @@
+  std::shared_ptr<Axle> get_axle(const std::string & axle_name);
+
+  // Handles for three wheels and their axles
+  std::shared_ptr<Wheel> front_left_wheel_handle_;
```
Great suggestion! I've updated the handle declarations to use std::unique_ptr since shared ownership isn't needed here. This should help make ownership clearer and reduce overhead.

nitin2606 committed:
Update swerve_drive_controller/src/swerve_drive_controller.cpp

Co-authored-by: Zheng Qu <quzhengrobot@gmail.com>

nitin2606 committed:
Changed Wheel and Axle handles to unique_ptr

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,740 @@
+  }
+  try {
+    // Declare parameters
+    auto_declare<std::string>("joint_steering_left_front", front_left_axle_joint_name_);
```
Sure, I'll do it.

nitin2606 committed:
Code Refactoring

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,192 @@
+  std::shared_ptr<Wheel> front_left_wheel_handle_;
+  std::shared_ptr<Wheel> front_right_wheel_handle_;
+  std::shared_ptr<Wheel> rear_left_wheel_handle_;
+  std::shared_ptr<Wheel> rear_right_wheel_handle_;
```
Seems good to put handles in vector

nitin2606 committed:
refactor(swervecontroller): optimize SwerveController code structure and performance

Improved code quality, reduced redundancy, and enhanced maintainability:
- Removed blocking sleep from on_configure lifecycle callback
- Ensured all wheel and axle interface accesses are null-checked
- Removed redundant set_velocity/set_position initializations
- Simplified initialization and cleanup logic
- Improved readability with consistent naming and formatting

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Resolved conflict in documentation

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

bmagyar committed:
Merge branch 'master' into feature/swerve-drive-controller

bmagyar committed:
Fix format

nitin2606 committed:
Mofied Joint And Axle Name Variables

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Modified CMakeLists.txt

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

Amronos commented on doc/migration.rst:
```diff
@@ -15,6 +15,4 @@ diff_drive_controller
-* The legacy ``antiwindup`` boolean and integral clamp parameters ``i_clamp_max``/``i_clamp_min`` have
-  been deprecated in favor of the new ``antiwindup_strategy`` parameter (`#1585 <https://github.com/ros-controls/ros2_controllers/pull/1585>`__). Choose a suitable anti-windup strategy and set the parameters accordingly.
-* PID state publisher topic changed to ``<controller_name>`` namespace and is initially turned off. It can be turned on by using  ``activate_state_publisher`` parameter. (`#1823 <https://github.com/ros-controls/ros2_controllers/pull/1823>`_).
+* The legacy ``antiwindup`` boolean and integral clamp parameters ``i_clamp_max``/``i_clamp_min`` have been deprecated in favor of the new ``antiwindup_strategy`` parameter (`#1585 <https://github.com/ros-controls/ros2_controllers/pull/1585>`__). Choose a suitable anti-windup strategy and set the parameters accordingly.
```
Looks like an incorrect merge, please undo this change.

Amronos commented on doc/release_notes.rst:
```diff
@@ -11,6 +11,8 @@ force_torque_sensor_broadcaster
 joint_trajectory_controller
 *******************************
 * The controller now supports the new anti-windup strategy of the PID class, which allows for more flexible control of the anti-windup behavior. (`#1759 <https://github.com/ros-controls/ros2_controllers/pull/1759>`__).
+
+=======
```
Looks like an incorrect merge, please undo this change.

nitin2606 replied:
done

Amronos commented on swerve_drive_controller/src/swerve_drive_controller_parameter.yaml:
Add ``read_only`` to the parameters.

Amronos commented on swerve_drive_controller/package.xml:
```diff
@@ -0,0 +1,39 @@
+  <name>swerve_drive_controller</name>
+  <version>5.0.0</version>
+  <description>Controller for a 4 wheel swerve drive robot base.</description>
+  <maintainer email="nitinmaurya2606@gmail.com">Nitin Maurya</maintainer>
```
Also add the other ros2_control maintainers here.

nitin2606 replied:
Will do it

Amronos commented on swerve_drive_controller/package.xml:
```diff
@@ -0,0 +1,39 @@
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>swerve_drive_controller</name>
+  <version>5.0.0</version>
```
```suggestion
  <version>0.0.0</version>
```
The versioning will be done by the maintainers.

nitin2606 replied:
Got it

Amronos commented on swerve_drive_controller/test/config/test_swerve_drive_controller.yaml:
```diff
@@ -0,0 +1,45 @@
+
+    cmd_vel_timeout: 10.0
+    use_stamped_vel: false
+    cmd_vel_topic: "/cmd_vel"
```
```suggestion
```

Amronos commented on swerve_drive_controller/src/swerve_drive_controller_parameter.yaml:
```diff
@@ -0,0 +1,83 @@
+      default_value: false,
+      description: "If true, velocity commands will be expected to have a timestamp."
+    }
+    cmd_vel_topic: {
+      type: string,
+      default_value: "/cmd_vel",
+      description: "Topic name for velocity commands."
+    }
```
This parameter is not needed; the topic can be remapped from the launch file.
```suggestion
```

Amronos commented on doc/release_notes.rst:
Add to the release notes that your controller was added.

nitin2606 replied:
Done

Amronos reviewed (changes requested):
I have done a rough review right now and haven't gone through the main source code properly.
Aside from my other review comments, I noticed the following:
1. You have the GPL parameters yaml, but aren't using GPL properly anywhere in your source code.
2. Why not make this controller a chained controller?
3. Are you using the new APIs here? (Look at #1566 and #1830)

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Mentioned swerve_driver_controller in release_notes.rst

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Made use of generate parameters library

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Removed unwanted merge

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Made parameters read only

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

Amronos commented on doc/migration.rst:
```diff
@@ -17,4 +17,4 @@ pid_controller
 * The legacy ``antiwindup`` boolean and integral clamp parameters ``i_clamp_max``/``i_clamp_min`` have
   been deprecated in favor of the new ``antiwindup_strategy`` parameter (`#1585 <https://github.com/ros-controls/ros2_controllers/pull/1585>`__). Choose a suitable anti-windup strategy and set the parameters accordingly.
-* PID state publisher topic changed to ``<controller_name>`` namespace and is initially turned off. It can be turned on by using  ``activate_state_publisher`` parameter. (`#1823 <https://github.com/ros-controls/ros2_controllers/pull/1823>`_).
+* PID state publisher topic changed to ``<controller_name>`` namespace and is initially turned off. It can be turned on by using  ``activate_state_publisher`` parameter. (`#1823 <https://github.com/ros-controls/ros2_controllers/pull/1823>`_).
```
Fix merge.
```suggestion
* PID state publisher topic changed to ``<controller_name>`` namespace and is initially turned off. It can be turned on by using  ``activate_state_publisher`` parameter. (`#1823 <https://github.com/ros-controls/ros2_controllers/pull/1823>`_).

```

Amronos commented on doc/release_notes.rst:
```diff
@@ -11,6 +11,7 @@ force_torque_sensor_broadcaster
 joint_trajectory_controller
 *******************************
 * The controller now supports the new anti-windup strategy of the PID class, which allows for more flexible control of the anti-windup behavior. (`#1759 <https://github.com/ros-controls/ros2_controllers/pull/1759>`__).
+
```
Fix merge
```suggestion
```

Amronos commented on swerve_drive_controller/package.xml:
```diff
@@ -0,0 +1,40 @@
+  <name>swerve_drive_controller</name>
+  <version>0.0.0</version>
+  <description>Controller for a 4 wheel swerve drive robot base.</description>
+  <maintainer email="nitinmaurya2606@gmail.com">Nitin Maurya</maintainer>
+  <maintainer email="christophfroehlich@users.noreply.github.com">Christoph Fröhlich</maintainer>
```
```suggestion
  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
  <maintainer email="christoph.froehlich@ait.ac.at">Christoph Froehlich</maintainer>
  <maintainer email="denis@stoglrobotics.de">Denis Štogl</maintainer>
  <maintainer email="sai.kishor@pal-robotics.com">Sai Kishor Kothakota</maintainer>
  <maintainer email="nitinmaurya2606@gmail.com">Nitin Maurya</maintainer>

  <author email="nitinmaurya2606@gmail.com">Nitin Maurya</author>
```

Amronos reviewed (changes requested):
Please add my suggestions to batch on the files changed tab and commit from there. This will make it easier to review.

nitin2606 committed:
Update doc/migration.rst

Co-authored-by: Aarav Gupta <aarav@spikonado.com>

nitin2606 committed:
Update doc/release_notes.rst

Co-authored-by: Aarav Gupta <aarav@spikonado.com>

nitin2606 committed:
Update swerve_drive_controller/package.xml

Co-authored-by: Aarav Gupta <aarav@spikonado.com>

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 commented:
@Amronos  @christophfroehlich  Could you please review when you get a chance?

christophfroehlich reviewed (changes requested):
To emphasize the last commit above: Please either use the github UI to accept proposed changes, or leave the conversations open so that the reviewing person can immediately check if they have been addressed or a new comment was added. Only the reviewer should manually close the conversations. This improves effectivity of the next review round.

Please fix the failing linter jobs: Install pre-commit via `pre-commit install` for the next commits; the current issues can be fixed with `pre-commit run --all` from the repo.

christophfroehlich commented on swerve_drive_controller/doc/userdoc.rst:
```diff
@@ -0,0 +1,47 @@
+Subscribers
+,,,,,,,,,,,,
+
+If ``use_stamped_vel=true``:
+~/cmd_vel [geometry_msgs/msg/TwistStamped]  
+  Velocity command for the controller. The controller extracts the x and y component of the linear velocity and the z component of the angular velocity. Velocities on other components are ignored.
+
+If ``use_stamped_vel=false``:
+~/cmd_vel [geometry_msgs/msg/Twist]  
+  Velocity command for the controller. The controller extracts the x and y component of the linear velocity and the z component of the angular velocity. Velocities on other components are ignored.
```
https://github.com/ros-controls/ros2_controllers/actions/runs/16826706501/job/47664758345?pr=1694
```
/home/runner/work/ros2_controllers/ros2_controllers/doc/ros2_controllers/swerve_drive_controller/doc/userdoc.rst:28: ERROR: Unexpected indentation.
/home/runner/work/ros2_controllers/ros2_controllers/doc/ros2_controllers/swerve_drive_controller/doc/userdoc.rst:32: ERROR: Unexpected indentation.
```

```suggestion
If ``use_stamped_vel=true``:
``~/cmd_vel [geometry_msgs/msg/TwistStamped]``
Velocity command for the controller. The controller extracts the x and y component of the linear velocity and the z component of the angular velocity. Velocities on other components are ignored.

If ``use_stamped_vel=false``:
``~/cmd_vel [geometry_msgs/msg/Twist]``
Velocity command for the controller. The controller extracts the x and y component of the linear velocity and the z component of the angular velocity. Velocities on other components are ignored.
```

something like this should fix the docs.

nitin2606 committed:
Update swerve_drive_controller/doc/userdoc.rst

Co-authored-by: Christoph Fröhlich <christophfroehlich@users.noreply.github.com>

nitin2606 committed:
Fixed Failing Linter Jobs

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 commented:
> To emphasize the last commit above: Please either use the github UI to accept proposed changes, or leave the conversations open so that the reviewing person can immediately check if they have been addressed or a new comment was added. Only the reviewer should manually close the conversations. This improves effectivity of the next review round.
> 
> Please fix the failing linter jobs: Install pre-commit via `pre-commit install` for the next commits; the current issues can be fixed with `pre-commit run --all` from the repo.


I have attempted to resolve the linter issues, but I am still encountering a persistent error reported by the pre-commit check at line 152 of the file:  https://github.com/nitin2606/ros2_controllers/blob/feature/swerve-drive-controller/swerve_drive_controller/test/test_swerve_drive_controller.cpp 
Could you kindly review it and help me identify the underlying issue?

christophfroehlich reviewed (changes requested):
A first review round, haven't gone over the kinematics implementation yet.

nitin2606 committed:
Update swerve_drive_controller/test/test_swerve_drive_controller.hpp

Co-authored-by: Christoph Fröhlich <christophfroehlich@users.noreply.github.com>

nitin2606 committed:
Update swerve_drive_controller/CMakeLists.txt

Co-authored-by: Christoph Fröhlich <christophfroehlich@users.noreply.github.com>

nitin2606 committed:
Update swerve_drive_controller/CMakeLists.txt

Co-authored-by: Christoph Fröhlich <christophfroehlich@users.noreply.github.com>

wsxq2 reviewed (changes requested):
I find a problem.

nitin2606 committed:
Modified Publisher Api, Test Files

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 commented:
@christophfroehlich, could you kindly review the changes in [test_swerve_drive_controller.cpp](https://github.com/nitin2606/ros2_controllers/blob/feature/swerve-drive-controller/swerve_drive_controller/test/test_swerve_drive_controller.cpp) While the code appears to be correct, it is still failing the pre-commit checks. I would also greatly appreciate it if you could review the other changes you suggested.

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

Narukara commented:
Thank you for all your work! I'm trying to test this controller locally with the Gazebo simulator. I'm searching for information on how to compile and use a new controller locally... 😃 If anyone could point me to a suitable tutorial, that would be helpful. I'll also try this myself.

nitin2606 commented:
> Thank you for all your work! I'm trying to test this controller locally with the Gazebo simulator. I'm searching for information on how to compile and use a new controller locally... 😃 If anyone could point me to a suitable tutorial, that would be helpful. I'll also try this myself.

I have made a local setup to test this controller, if you want I can share the github link.

Narukara commented:
> I have made a local setup to test this controller, if you want I can share the github link.

Yes, please! That would be extremely helpful. Thank you for offering to share it. :smile:

nitin2606 commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,530 @@
+  {-0.1, 0.175},   // front left
+  {0.1, 0.175},    // front right
+  {-0.1, -0.175},  // rear left
+  {0.1, -0.175}    // rear right
```
Yes you are right, wheel positions must be taken via parameters. Sorry for the late response. 
I'll make necessary changes and push.

Narukara replied:
I've noticed another issue. According to ROS REP 103, the forward direction of a vehicle should align with the X-axis. That means the positions of the four wheels in the coordinate frame should be:

```
            ^y
  Rear Left |  Front Left
------------+-------------->x
 Rear Right |  Front Right
```

This is also how the Ackermann vehicle model is designed in [this URDF file](https://github.com/ros-controls/gz_ros2_control/blob/rolling/gz_ros2_control_demos/urdf/test_ackermann_drive.xacro.urdf). However, the current placement here does not seem to comply with this, which could potentially have some implications for the code.

Narukara commented:
There's another issue: the `userdoc.rst` mentions that the `use_stamped_vel` option can be used to configure the controller to accept either `TwistStamped` or `Twist` messages. However, this does not appear to be implemented in the code — the controller always uses `TwistStamped`. :thinking:

Narukara commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,631 @@
+  double & linear_x_cmd = command.twist.linear.x;
+  double & linear_y_cmd = command.twist.linear.y;
+  double & angular_cmd = command.twist.angular.z;
+  double x_offset = wheel_params_.x_offset;
+  double y_offset = wheel_params_.y_offset;
+  double radius = wheel_params_.radius;
```
I did some tests in Gazebo: using a car with a wheel radius of 0.08m, sending a command of tw.linear.x = 0.5 to the controller, the actual speed of the car was about 0.04m/s. This suggests that the controller is directly using the speed command as the angular velocity.

nitin2606 commented:
> There's another issue: the `userdoc.rst` mentions that the `use_stamped_vel` option can be used to configure the controller to accept either `TwistStamped` or `Twist` messages. However, this does not appear to be implemented in the code — the controller always uses `TwistStamped`. 🤔

userdoc.rst is not updated. All the controllers of ros2_control have switched to TwistStamped only.

nitin2606 committed:
Instead of hardcoding wheel positions, wheel coordinates are now
computed from wheelbase, trackwidth, offset_x, and offset_y. This
improves flexibility and reduces duplication when changing robot
dimensions.

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

Narukara commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,523 @@
+    }
+    else
+    {
+      velocity_array[i] = wheel_handles_[i]->get_feedback();
```
```suggestion
      velocity_array[i] = wheel_handles_[i]->get_feedback() * params_.wheel_radius;
```

nitin2606 replied:
Solved in latest commit

Narukara reviewed (commented):
I'm not sure if this is the right place to handle the wheel radius. I tested the modified controller using gazebo, and both the controls and odometry seem to work fine. By the way, I used the humble version, so I made some necessary changes.

Narukara commented:
Another small issue: I found that when I stop sending speed commands, the controller automatically straightens all the wheels. Maybe it would be more reasonable to keep each wheel at its current steering angle?

nitin2606 committed:
refactor(swerve): improve wheel kinematics and controller behavior

 - Updated kinematics to compute angular velocity for wheels instead
   of linear velocity for more accurate motion control.
 - Enhanced controller to retain the last steering angle when stopped,
   preventing the wheels from snapping back to zero orientation.

nitin2606 commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,523 @@
+        wheel_joint_names[i]);
+    }
+    axle_handles_[i]->set_position(wheel_command[i].steering_angle);
+    wheel_handles_[i]->set_velocity(wheel_command[i].drive_velocity);
```
Solved in latest commit

nitin2606 commented:
> Another small issue: I found that when I stop sending speed commands, the controller automatically straightens all the wheels. Maybe it would be more reasonable to keep each wheel at its current steering angle?

please check latest commit

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

Narukara commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,539 @@
+  tf2::Quaternion orientation;
+  orientation.setRPY(0.0, 0.0, odometry_.theta);
+
+  if (realtime_odometry_publisher_)
+  {
+    auto & odometry_message = realtime_odometry_publisher_->msg_;
+    odometry_message.header.stamp = time;
+    odometry_message.pose.pose.position.x = odometry_.x;
+    odometry_message.pose.pose.position.y = odometry_.y;
+    odometry_message.pose.pose.orientation.x = orientation.x();
+    odometry_message.pose.pose.orientation.y = orientation.y();
+    odometry_message.pose.pose.orientation.z = orientation.z();
+    odometry_message.pose.pose.orientation.w = orientation.w();
+    realtime_odometry_publisher_->tryPublish(odometry_message);
+  }
```
The odometry message published here does not contain twist information (twist.linear.x, twist.linear.y, twist.angular.z). I have confirmed this issue in the Gazebo test.

Narukara reviewed (changes requested):
add odometry twist

nitin2606 committed:
Update swerve_drive_controller/include/swerve_drive_controller/swerve_drive_kinematics.hpp

Co-authored-by: Narukara <narukara17@gmail.com>

nitin2606 committed:
Update swerve_drive_controller/src/swerve_drive_controller.cpp

Co-authored-by: Narukara <narukara17@gmail.com>

nitin2606 committed:
Update swerve_drive_controller/src/swerve_drive_kinematics.cpp

Co-authored-by: Narukara <narukara17@gmail.com>

nitin2606 committed:
Update swerve_drive_controller/src/swerve_drive_kinematics.cpp

Co-authored-by: Narukara <narukara17@gmail.com>

nitin2606 committed:
Made swerve_drive_controller chainable

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
Suppressed false syntax error

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

nitin2606 committed:
choose minimal steering rotation and reverse drive when needed

Signed-off-by: nitin <nitinmaurya2606@gmail.com>

Narukara commented:
> I think this is slightly problematic as it this situation can occur very quickly with steering motors of a limited range.

The control of a four-wheel independent steering vehicle with **steering angle limitations** might be a rather complex problem. Personally, I suggest that we only complete the basic functions in this PR and leave this problem to the next issue/PR. 🙂

tgaspar commented:
Thank you @Narukara for the feedback.

The link to the specifications/documentation of the steering mechanism regarding the Z axis is very useful. I have a similar definition of the steering joint as yours, it's just that its parent frame Z axis points downwards :slightly_smiling_face: Subsequently also the steering axis points down.

Given that this is an isolated issue, I understand that it does not need to make it into the PR.

As for the limited rotation of the steering axes: totally agree! If it's a complex problem to solve it can be adjusted iteratively later. In this case, I would  find it reasonable to note this limitation somewhere.

nitin2606 committed:
Update swerve_drive_controller/src/swerve_drive_controller.cpp

Co-authored-by: Narukara <narukara17@gmail.com>

nitin2606 commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,624 @@
+    for (std::size_t i = 0; i < 6; ++i)
+    {
+      params_.twist_covariance_diagonal[i] = 0.01;
+    }
```
Yes, it should be read from configs. I'll make changes

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
feat(odometry): improve odometry accuracy with exact integration
Replace Euler integration with Exact + RK2 integration for better
odometry accuracy during combined translation and rotation maneuvers.
- Use Exact integration when angular velocity is significant
- Fall back to RK2 for near-zero angular velocity

Signed-off-by: nitin2606 <nitinmaurya2606@gmail.com>

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
fix: removed hardcoded covariance, set sensible defaults in params

Signed-off-by: nitin2606 <nitinmaurya2606@gmail.com>

nitin2606 committed:
Fix: Revert doc changes (already merged in PR #1712) to fix CI duplicate errors

Signed-off-by: nitin2606 <nitinmaurya2606@gmail.com>

nitin2606 committed:
added swerve_drive_controller in controller_index.rst

Signed-off-by: nitin2606 <nitinmaurya2606@gmail.com>

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

nitin2606 committed:
Merge branch 'master' into feature/swerve-drive-controller

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+  auto logger = get_node()->get_logger();
+
+  wheel_handles_.resize(4);
+  for (std::size_t i = 0; i < 4; i++)
```
is `on_activate` really the best place for this?
looking at a handful of other controllers, it seems like this type of thing is generally put into `on_configure`

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+
+  for (std::size_t i = 0; i < wheel_handles_.size(); ++i)
+  {
+    if (!wheel_handles_[i])
```
is there a reason this (the validation part, not the resetting part) isn't in the same for loop that creates the wheel handles?

solonovamax commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,238 @@
+  OdometryState odometry_;
+
+  // Topic Subscription
+  bool subscriber_is_active_ = false;
```
it's not obvious to me why this is necessary?

other controllers seem to just blindly update the stored command value from the topic, even if deactivated, as there's not really any harm in that

solonovamax commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,238 @@
+  rclcpp::Time previous_publish_timestamp_{0, 0, RCL_CLOCK_UNINITIALIZED};
+
+  // Timeout to consider cmd_vel commands old
+  std::chrono::milliseconds cmd_vel_timeout_{500};
```
shouldn't this be an `rclcpp::Duration`?

solonovamax commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,238 @@
+  Params params_;
+
+  SwerveDriveKinematics swerveDriveKinematics_;
+  std::queue<TwistStamped> previous_commands_;  // last two commands
```
seems to be unused?

regardless, storing a queue for two elements probably isn't the best idea
you could just have two fields with pointers and then swap them before overwriting the newer one, that way you avoid additional allocations

solonovamax commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,238 @@
+
+protected:
+  // Handles for four wheels and their axles
+  std::vector<std::optional<Wheel>> wheel_handles_;
```
shouldn't these be `std::array`s as well, if there's only ever four wheels?

solonovamax replied:
are these optional exclusively for the error handling?

imo it would be better to just make these not optional & have `get_interface_object()` return an optional, then in `on_configure`/`on_activate` if they cannot be found just erroring out

it avoids an necessary pointer dereference during normal operation (because optional is just a wrapper around a raw pointer + nullptr)
it also avoids a bunch of otherwise unecessary `has_value()` calls, because those can never happen.

solonovamax commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,238 @@
+  TwistStamped command_msg_;
+  std::shared_ptr<rclcpp::Publisher<nav_msgs::msg::Odometry>> odometry_publisher_ = nullptr;
+  std::shared_ptr<realtime_tools::RealtimePublisher<nav_msgs::msg::Odometry>>
+    realtime_odometry_publisher_ = nullptr;
```
other controllers use a unique pointer for the realtime publisher

solonovamax commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,238 @@
+    realtime_odometry_transform_publisher_ = nullptr;
+  tf2_msgs::msg::TFMessage odometry_transform_message_;
+
+  bool is_halted_ = false;
```
this feels like it's reinventing the already existing lifecycles to me

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+      geometry_msgs::msg::TwistWithCovariance(rosidl_runtime_cpp::MessageInitialization::ALL);
+
+    constexpr std::size_t NUM_DIMENSIONS = 6;
+    for (std::size_t index = 0; index < 6; ++index)
```
`NUM_DIMENSIONS` should probably be used instead of `6` here

solonovamax commented on swerve_drive_controller/include/swerve_drive_controller/swerve_drive_controller.hpp:
```diff
@@ -0,0 +1,238 @@
+
+using CallbackReturn = controller_interface::CallbackReturn;
+
+class Wheel
```
tbh is it really the best to encapsulate these as wheel + axel separetely?

I haven't looked into this too much, however I feel it may make more sense to group them in something along the lines of a `SwerveDriveModule` which holds wheel & axes interfaces, and then you would have like

```cpp
set_command(double velocity, double rotation);

std::pair<double, double> get_state();
```

also probably a good idea to just make this a struct, as it avoids the vtable lookup (though the compiler might optimize that away, I'm not familiar enough with C++ to say)

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during init stage with message: %s \n", e.what());
```
the node is guaranteed to be available at this point, so `RCLCPP_ERROR` should be used instead

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+  auto node = get_node();
+  if (!node)
+  {
+    RCLCPP_ERROR(node->get_logger(), "Node is null in on_init");
```
the node does not exist, so `RCLCPP_ERROR` cannot be used because `node->get_logger()` is a null pointer dereference

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+    {wheel_command[3], params_.rear_right_velocity_threshold / params_.wheel_radius,
+     "rear_right_wheel"}};
+
+  for (const auto & [wheel_command_, threshold, label] : wheel_data)
```
does not account for the negative case (e.g. velocity is `-2.5`, threshold is `2`)

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+    }
+  }
+
+  const double min_steering_error = M_PI / 6.0;  // 30 degrees
```
why 30 degrees?

also, should probably be a `static constexpr`

solonovamax commented on swerve_drive_controller/src/swerve_drive_controller.cpp:
```diff
@@ -0,0 +1,615 @@
+    double velocity_scale = 1.0;
+    if (steering_error > min_steering_error)
+    {
+      if (steering_error >= 1.5608)  // ~89.5 degrees
```
might be better to use a ternary here (or, better yet, `std::min`)

also, why this value? why not 89.9 degrees? etc.

solonovamax commented on swerve_drive_controller/src/swerve_drive_kinematics.cpp:
```diff
@@ -0,0 +1,172 @@
+
+  if (wheel_radius <= 0.0)
+  {
+    std::cerr << "invalid wheel_radius <= 0.0\n";
```
should use the normal rclcpp logging

solonovamax commented on swerve_drive_controller/src/swerve_drive_kinematics.cpp:
```diff
@@ -0,0 +1,172 @@
+  if (wheel_radius <= 0.0)
+  {
+    std::cerr << "invalid wheel_radius <= 0.0\n";
+    // fallthrough: compute but set angular velocities to 0 to avoid div-by-zero
```
wouldn't it be better to just completely error out?

solonovamax commented on swerve_drive_controller/src/swerve_drive_kinematics.cpp:
```diff
@@ -0,0 +1,172 @@
+
+    wheel_commands[i].drive_velocity = linear_speed;
+
+    if (wheel_radius > 0.0)
```
probably better to use a ternary here

solonovamax commented on swerve_drive_controller/src/swerve_drive_kinematics.cpp:
```diff
@@ -0,0 +1,172 @@
+  double vx_sum = 0.0, vy_sum = 0.0, wz_sum = 0.0;
+  for (std::size_t i = 0; i < 4; i++)
+  {
+    double vx = wheel_velocities[i] * std::cos(steering_angles[i]);
```
unsure if this is a GNU extension or not, but in math.h there's `sincos` which calculates the sine & cosine values simultaneously (though, the compiler might also optimize this for you, unsure)

solonovamax reviewed (commented):
just sort of skimmed through this on my phone (that's why each comment is only on a single line, sorry about that) and had some random comments

not super familiar with how things are done in this codebase however, so take some things with a grain of salt
